pkg/controller/hibernation/hibernation_controller.go

```diff
@@ -224,29 +224,58 @@ func (r *hibernationReconciler) Reconcile(ctx context.Context, request reconcile
 		}
 	}
 
-	// Check if HibernateAfter is set, and if the cluster has been in running state for longer than this duration, put it to sleep.
-	if cd.Spec.HibernateAfter != nil && cd.Spec.PowerState != hivev1.HibernatingClusterPowerState {
+	// Check if HibernateAfter is set, and put it to sleep if necessary.
+	// As the baseline timestamp for determining whether HibernateAfter should trigger, we use the latest of:
+	// - When the cluster finished installing (status.installedTimestamp)
+	// - When the cluster was claimed (status.claimedTimestamp -- ClusterPool CDs only)
+	// - The last time the cluster resumed (status.conditions[Hibernating].lastTransitionTime if not hibernating (but see TODO))
+	// BUT pool clusters wait until they're claimed for HibernateAfter to have effect.
+	isUnclaimedPoolCluster := cd.Spec.ClusterPoolRef != nil &&
+		cd.Spec.ClusterPoolRef.PoolName != "" &&
+		// If we happen to hit this code path between the CD's spec and status update, the ClaimName
+		// will be set but the ClaimedTimestamp won't. We don't want to treat this as "claimed" just
+		// yet, because we rely on the timestamp being set.
+		(cd.Spec.ClusterPoolRef.ClaimName == "" || cd.Status.ClaimedTimestamp == nil)
+	if cd.Spec.HibernateAfter != nil && cd.Spec.PowerState != hivev1.HibernatingClusterPowerState && !isUnclaimedPoolCluster {
 		hibernateAfterDur := cd.Spec.HibernateAfter.Duration
-		runningSince := cd.Status.InstalledTimestamp.Time
-		hibLog := cdLog.WithFields(log.Fields{
-			"runningSince":   runningSince,
-			"hibernateAfter": hibernateAfterDur,
-		})
+		hibLog := cdLog.WithField("hibernateAfter", hibernateAfterDur)
+
+		// The nil values of each of these will make them "earliest" and therefore unused.
+		var installedSince, runningSince, claimedSince time.Time
 		var isRunning bool
 
+		installedSince = cd.Status.InstalledTimestamp.Time
+		hibLog = hibLog.WithField("installedSince", installedSince)
+
+		if cd.Status.ClaimedTimestamp != nil {
+			claimedSince = cd.Status.ClaimedTimestamp.Time
+			hibLog = hibLog.WithField("claimedSince", claimedSince)
+		} else {
+			// This means it's not a pool cluster (!isUnclaimedPoolCluster && ClaimedTimestamp == nil)
+			hibLog.Debug("cluster does not belong to a clusterpool")
+		}
+
 		cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.ClusterHibernatingCondition)
 		if cond.Status == corev1.ConditionUnknown {
-			hibLog.Debug("cluster has never been hibernated, using installed time")
+			hibLog.Debug("cluster has never been hibernated")
```

Comment:
Two lines above, it seems like `cond` can potentially be nil. Although I understand that we set conditions to default values, can we defensively code this to avoid inadvertent panics? Or is there some mechanism that is guaranteeing that this simply cannot happen?

Reply:
Legit. Fixed.